flutter/lib/src/span_frame_metrics_collector.dart

```diff
@@ -0,0 +1,226 @@
+
+    final frameMetrics =
+        calculateFrameMetrics(span, endTimestamp, displayRefreshRate);
+    _applyFrameMetricsToSpan(span, frameMetrics);
+
+    activeSpans.remove(span);
+    if (activeSpans.isEmpty) {
+      clear();
+    } else {
+      frameDurations.removeWhere(
+          (key, _) => key.isBefore(activeSpans.first.startTimestamp));
+    }
+  }
+
+  /// Calls [WidgetsBinding.instance.addPersistentFrameCallback] which cannot be unregistered
+  /// and exists for the duration of the application's lifetime.
+  ///
+  /// Stopping the frame tracking means setting [isTrackingPaused] is `true`
+  /// to prevent actions being done when the frame callback is triggered.
+  void startFrameTracking() {
+    _isTrackingPaused = false;
+
+    if (!_isTrackingRegistered) {
+      _frameCallbackHandler.addPersistentFrameCallback(measureFrameDuration);
+      _isTrackingRegistered = true;
+    }
+  }
+
+  /// Records the duration of a single frame and stores it in [frameDurations].
+  ///
+  /// This method is called for each frame when frame tracking is active.
+  Future<void> measureFrameDuration(Duration duration) async {
+    // Using the stopwatch to measure the frame duration is flaky in ci
+    if (_isTestMode) {
+      // ignore: invalid_use_of_internal_member
+      frameDurations[options.clock()] = duration.inMilliseconds;
+      return;
+    }
+
+    if (_isTrackingPaused) return;
+
+    if (!_stopwatch.isRunning) {
+      _stopwatch.start();
+    }
+
+    await _frameCallbackHandler.endOfFrame;
+
+    final frameDuration = _stopwatch.elapsedMilliseconds;
+    // ignore: invalid_use_of_internal_member
+    frameDurations[options.clock()] = frameDuration;
+
+    _stopwatch.reset();
+
+    if (_frameCallbackHandler.hasScheduledFrame == true) {
+      _stopwatch.start();
+    }
+  }
+
+  void _applyFrameMetricsToSpan(
+      ISentrySpan span, Map<String, int> frameMetrics) {
+    frameMetrics.forEach((key, value) {
+      span.setData(key, value);
+    });
+
+    // This will call the methods on the tracer, not on the span directly
+    if (span is SentrySpan && span.isRootSpan) {
+      frameMetrics.forEach((key, value) {
+        // ignore: invalid_use_of_internal_member
+        span.tracer.setData(key, value);
+      });
+      frameMetrics.forEach((key, value) {
+        // In measurements we change e.g frames.total to frames_total
+        // We don't do span.tracer.setMeasurement because setMeasurement in SentrySpan
+        // uses the tracer internally
+        span.setMeasurement(key.replaceAll('.', '_'), value);
+      });
+    }
+  }
+
+  @visibleForTesting
+  Map<String, int> calculateFrameMetrics(
+      ISentrySpan span, DateTime endTimestamp, int displayRefreshRate) {
+    final expectedFrameDuration = ((1 / displayRefreshRate) * 1000).toInt();
+
+    // Filter frame durations within the span's time range
+    final timestamps = frameDurations.keys
+        .takeWhile((value) =>
```

Comment:
sadly `takeWhile` doesn't work here.
If a span is finished and there is a frame in the map before the start of the span, this takeWhile will return an empty collection. 
The problem is `value.isAfter(span.startTimestamp)`
e.g.
span a -> start: 1, end: 10
span b -> start: 3, end: 8
when span  b finishes, there is a frame at t 1 and 2, so the condition of the takeWhile will be false

Reply:
I've updated the implementation that also includes partially contained frames similar to android, hopefully it makes sense